google/cloud/bigtable/internal/async_read_stream_impl.h

```diff
@@ -0,0 +1,350 @@
+ * a `shared_ptr`. This is enforced by making the constructor private and the
+ * one factory function returns a `shared_ptr`.
+ *
+ * @tparam Response the type of the responses in the streaming read RPC.
+ * @tparam OnReadHandler the type of the user-provided callable to handle Read
+ *     responses.
+ * @tparam OnFinishHandler the type of the user-provided callback to handle
+ *     Finish responses.
+ */
+template <typename Response, typename OnReadHandler, typename OnFinishHandler>
+class AsyncReadStreamImpl
+    : public AsyncOperation,
+      public std::enable_shared_from_this<
+          AsyncReadStreamImpl<Response, OnReadHandler, OnFinishHandler>> {
+ public:
+  /**
+   * Create a new instance.
+   *
+   * @param on_read the handler for a successful `Read()` result. Failed
+   *   `Read()` operations automatically terminate the loop and call `Finish()`.
+   * @param on_finish the handler for a completed `Finish()` result.
+   */
+  static std::shared_ptr<AsyncReadStreamImpl> Create(
+      OnReadHandler&& on_read, OnFinishHandler&& on_finish) {
+    return std::shared_ptr<AsyncReadStreamImpl>(
+        new AsyncReadStreamImpl(std::forward<OnReadHandler>(on_read),
+                                std::forward<OnFinishHandler>(on_finish)));
+  }
+
+  /**
+   * Start the asynchronous streaming read request and its read loop.
+   *
+   * @param async_call the function that will make the asynchronous
+   *     streaming read RPC.  This is typically a wrapper around one of the
+   *     gRPC-generated `PrepareAsync*()` functions.
+   * @param request the request parameter for the streaming read RPC.
+   * @param context the client context to control the streaming read RPC.
+   * @param cq the completion queue that will execute the streaming read RPC. It
+   *     is the application's responsibility to keep a thread pool to execute
+   *     the completion queue loop.
+   *
+   * @tparam AsyncFunctionType the type for @p async_call.
+   * @tparam Request the type for @p request.
+   *
+   */
+  template <typename AsyncFunctionType, typename Request>
+  void Start(AsyncFunctionType&& async_call, Request const& request,
+             std::unique_ptr<grpc::ClientContext> context,
+             std::shared_ptr<CompletionQueueImpl> cq) {
+    // An adapter to call OnStart() via the completion queue.
+    class NotifyStart final : public AsyncGrpcOperation {
+     public:
+      explicit NotifyStart(std::shared_ptr<AsyncReadStreamImpl> c)
+          : control_(std::move(c)) {}
+
+     private:
+      void Cancel() override {}  // LCOV_EXCL_LINE
+      bool Notify(CompletionQueue&, bool ok) override {
+        control_->OnStart(ok);
+        return true;
+      }
+      std::shared_ptr<AsyncReadStreamImpl> control_;
+    };
+
+    context_ = std::move(context);
+    cq_ = std::move(cq);
+    reader_ = async_call(context_.get(), request, &cq_->cq());
+    auto self = this->shared_from_this();
+    auto callback = std::make_shared<NotifyStart>(self);
```

Comment:
std::move(self)?

Reply:
Done (well, something similar, let me know what you think).